src/PrestaShopBundle/Controller/Admin/Improve/International/CurrencyController.php

```diff
@@ -205,6 +253,135 @@ public function toggleStatusAction($currencyId)
         return $this->redirectToRoute('admin_currencies_index');
     }
 
+    /**
+     * Updates exchange rates.
+     *
+     * @AdminSecurity(
+     *     "is_granted('update', request.get('_legacy_controller'))",
+     *     redirectRoute="admin_currencies_index",
+     *     message="You do not have permission to edit this."
+     * )
+     * @DemoRestricted(redirectRoute="admin_currencies_index")
+     *
+     * @return RedirectResponse
+     */
+    public function updateExchangeRatesAction()
+    {
+        try {
+            $this->getCommandBus()->handle(new RefreshExchangeRatesCommand());
+
+            $this->addFlash('success', $this->trans('Successful update.', 'Admin.Notifications.Success'));
+        } catch (CannotRefreshExchangeRatesException $exception) {
+            $this->addFlash('error', $exception->getMessage());
+        }
+
+        return $this->redirectToRoute('admin_currencies_index');
+    }
+
+    /**
+     * Handles ajax request which updates live exchange rates.
+     *
+     * @param Request $request
+     *
+     * @return JsonResponse
+     */
+    public function updateLiveExchangeRatesAction(Request $request)
+    {
+        if ($this->isDemoModeEnabled()) {
+            return $this->json([
+                    'status' => false,
+                    'message' => $this->getDemoModeErrorMessage(),
+                ],
+                Response::HTTP_UNAUTHORIZED
+            );
+        }
+
+        $authLevel = $this->authorizationLevel($request->attributes->get('_legacy_controller'));
```

Comment:
However I think you can use `@AdminSecurity` even for ajax routes. Maybe you did not want to do it because you did not want to redirect ?

Again, maybe there is room for improvement. If you do not want to make this PR bigger, tell me and we'll make it the subject for another PR 😄

Reply:
For @AjaxDemoRestricted and @AdminSecurity for ajax I though about we can do something like that - but it seems like totally different functionality so I would rather introduce it in another PR to keep things simple and it would be way easier for someone to review it ;) Should I add reminder for myself or maybe we should raise an issue with it for improvement? :thinking:

Reply:
Done 😄 https://github.com/PrestaShop/PrestaShop/issues/12083